design/one-pager-default-resource-class.md

```diff
@@ -88,66 +116,237 @@ spec:
+   embeddable struct
+   [`Policy`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L93),
+   then bumping the dependency in core Crossplane.
+1. Remove the `classRef` field from all claims entirely. This would require the
+   same steps as above but with
+   [`ResourceClaimSpec`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L49).
+1. Add a `providerRef` field to all claims using the same method detailed above.
+   This would require updating the same steps as above also with
+   [`ResourceClaimSpec`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L49).
+1. Alter all claim controllers (which live in provider stacks) to only reconcile
+   claims that have a `policyRef` to a `Policy` that has a `classRef` of that
+   kind's strongly-typed class. For instance, the `RDSInstance` claim controller
+   will only reconcile `MySQLInstance` claims that have `policyRef` to a
+   `MySQLInstancePolicy` that has a `classRef` to an `RDSInstanceClass`. This
+   will involve updating the claim controller
+   [predicates](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/pkg/resource/predicates.go#L41),
+   as well as the actual logic of the [shared claim
+   reconciler](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/pkg/resource/claim_reconciler.go#L281),
+   both in `crossplane-runtime`. It should require minimal, if any, updates to
+   the actual claim controllers in each of the provider stacks.
+
+In this model, a claim will now look as follows:
 
 ```yaml
-apiVersion: storage.crossplane.io/v1alpha1
-kind: NoSQLInstance
+apiVersion: database.crossplane.io/v1alpha1
+kind: MySQLInstance
 metadata:
-  name: dynamo-claim
-  namespace: demo
+  name: mysql-claim
 spec:
-  ...
+  policyRef:
+    name: standard-mysql
+    namespace: crossplane-system
+  writeConnectionSecretToRef:
+    name: rdsmysql
+  engineVersion: "5.6"
 ```
 
-Or in the more specific case:
+With an arbitrary number of potential `MySQLInstancePolicy` objects that could
+be referenced:
 
 ```yaml
-apiVersion: storage.crossplane.io/v1alpha1
-kind: DynamoDBInstance
+apiVersion: database.crossplane.io/v1alpha1
+kind: MySQLInstancePolicy
 metadata:
-  name: dynamo-claim
-  namespace: demo
-spec:
-  ...
+  name: standard-mysql
+  namespace: crossplane-system
+classRef:
+  kind: RDSInstanceClass
+  apiVersion: database.aws.crossplane.io/v1alpha1
+  name: standard-mysql
+  namespace: crossplane-system
 ```
 
+### Denote Default via Label
+
+This feature is similar to the original default class model in that it uses
+labels to specify which class of service to use as default when a `policyRef` is
+omitted. For each `namespace`, there must be only one policy that has the
```

Comment:
> there must be only one policy

How do we expect to enforce this? Is there a way to enforce this at the time the policy is created?

Related: what is the reason for this restriction? What if I want to round-robin my created resources between multiple configurations? This is sounding more like a Best Practice than a Hard Requirement to me.

Reply:
Right now this requirement is because we would have no way of knowing which policy to default to if multiple were defined as `default`. I definitely see a more intelligent and varied defaulting pattern in the future. The first iteration could involve a `weight` label or something of that nature where the highest weighted policy in the namespaces could be applied. However, I am not sure that this is in scope for this iteration. Would love to hear other thoughts on this.

cc @prasek @negz

Reply:
The only way I'm aware to enforce this would be to use an admission webhook (or rather a series thereof) that rejected new policies that were labelled `default` for any namespace that already contained the same kind of policy labelled such. Crossplane currently has no webhooks, and likely won't for 0.3, so at least at first we'd probably just have to make it clear that the default policy is undefined when multiple policies are labelled `default`.